From dbd8ffb54f72f45acc644f1f16709dd549d0cfc9 Mon Sep 17 00:00:00 2001
From: Luc Maisonobe <luc@apache.org>
Date: Fri, 26 Sep 2008 19:51:13 +0000
Subject: [PATCH] propagated trunk changes from r699157 into branch 2.0
 MATH-227.  fixed F distribution inverse CDF computation for small denominator
 degrees of freedom.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/math/branches/MATH_2_0@699468 13f79535-47bb-0310-9956-ffa450edef68
---
 .../math/distribution/FDistributionImpl.java      |  9 +++++++--
 src/site/xdoc/changes.xml                         |  3 +++
 .../math/distribution/FDistributionTest.java      | 15 +++++++++++++++
 3 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/src/java/org/apache/commons/math/distribution/FDistributionImpl.java b/src/java/org/apache/commons/math/distribution/FDistributionImpl.java
index 3959403659..59aeb07b03 100644
--- a/src/java/org/apache/commons/math/distribution/FDistributionImpl.java
+++ b/src/java/org/apache/commons/math/distribution/FDistributionImpl.java
@@ -141,8 +141,13 @@ protected double getDomainUpperBound(double p) {
      * @return initial domain value
      */
     protected double getInitialDomain(double p) {
-        return getDenominatorDegreesOfFreedom() /
-            (getDenominatorDegreesOfFreedom() - 2.0);
+        double ret = 1.0;
+        double d = getDenominatorDegreesOfFreedom();
+        if (d > 2.0) {
+            // use mean
+            ret = d / (d - 2.0);
+        }
+        return ret;
     }
     
     /**
diff --git a/src/site/xdoc/changes.xml b/src/site/xdoc/changes.xml
index b626d7d0e2..69897d590e 100644
--- a/src/site/xdoc/changes.xml
+++ b/src/site/xdoc/changes.xml
@@ -39,6 +39,9 @@ The <action> type attribute can be add,update,fix,remove.
   </properties>
   <body>
     <release version="2.0" date="TBD" description="TBD">
+      <action dev="brentworden" type="fix" issue="MATH-227" due-to="Joerg Henning">
+        Fixed F distribution inverse CDF computation for small denominator degrees of freedom.
+      </action>
       <action dev="luc" type="fix" issue="MATH-226" due-to="Stuart Siegel">
         Fixed an error in CorrelatedRandomVectorGenerator leading to a component of
         the generated vector being constant.
diff --git a/src/test/org/apache/commons/math/distribution/FDistributionTest.java b/src/test/org/apache/commons/math/distribution/FDistributionTest.java
index d345d68e90..de153e0064 100644
--- a/src/test/org/apache/commons/math/distribution/FDistributionTest.java
+++ b/src/test/org/apache/commons/math/distribution/FDistributionTest.java
@@ -105,4 +105,19 @@ public void testLargeDegreesOfFreedom() throws Exception {
         double x = fd.inverseCumulativeProbability(p);
         assertEquals(.999, x, 1.0e-5);
     }
+
+    public void testSmallDegreesOfFreedom() throws Exception {
+        org.apache.commons.math.distribution.FDistributionImpl fd =
+            new org.apache.commons.math.distribution.FDistributionImpl(
+                1.0, 1.0);
+        double p = fd.cumulativeProbability(0.975);
+        double x = fd.inverseCumulativeProbability(p);
+        assertEquals(0.975, x, 1.0e-5);
+
+        fd.setDenominatorDegreesOfFreedom(2.0);
+        p = fd.cumulativeProbability(0.975);
+        x = fd.inverseCumulativeProbability(p);
+        assertEquals(0.975, x, 1.0e-5);
+    }
+
 }
